File: src/VCFProgenyPhased.cpp
#include <stack>
#include <sstream>
#include <algorithm>
#include <cmath>
#include <random>
#include <cassert>
#include "../include/common.h"
#include "../include/VCFProgenyPhased.h"
#include "../include/VCFFillable.h"
#include "../include/Map.h"
#include "../include/Pedigree.h"
#include "../include/KnownFamily.h"
#include "../include/Imputer.h"

using namespace std;


//////////////////// VCFProgenyPhased ////////////////////

VCFProgenyPhased::VCFProgenyPhased(const vector<STRVEC>& h, const STRVEC& s,
									vector<VCFFamilyRecord *> rs,
									const Map& gmap, const VCFSmall *ref) :
				VCFBase(h, s), VCFFamilyBase(), VCFImputable(gmap),
				records(rs), selection(0), ref_vcf(ref) { }

VCFProgenyPhased::~VCFProgenyPhased() {
	for(auto p = records.begin(); p != records.end(); ++p)
		delete *p;
}

Haplotype VCFProgenyPhased::clip_ref_haplotype(
								size_t sample_index, int side) const {
	const vector<int>	hap = ref_vcf->clip_raw_haplotype(sample_index, side);
	return Haplotype(hap, sample_index, side);
}

VCFProgenyPhased *VCFProgenyPhased::divide_by_positions(
											size_t first, size_t last) const {
	vector<VCFFamilyRecord *>	rs(records.begin() + first,
								   records.begin() + last);
	vector<VCFRecord *>	ref_rs(ref_vcf->get_records().begin() + first,
							   ref_vcf->get_records().begin() + last);
	const VCFSmall	*ref_vcf_divided = new VCFSmall(ref_vcf->get_header(),
													ref_vcf->get_samples(),
													ref_rs, true);
	return new VCFProgenyPhased(header, samples, rs, get_map(),
														ref_vcf_divided);
}

vector<Haplotype> VCFProgenyPhased::collect_haplotypes_from_phased_progeny(
															int side) const {
	vector<Haplotype>	haps(1, clip_haplotype(2, side));
	return haps;
}

vector<Haplotype> VCFProgenyPhased::collect_haplotype_from_refs() const {
	vector<Haplotype>	haps;
	for(size_t sample_index = 0; sample_index < ref_vcf->num_samples();
														++sample_index) {
		haps.push_back(clip_ref_haplotype(sample_index, 0));
		haps.push_back(clip_ref_haplotype(sample_index, 1));
	}
	return haps;
}

vector<Haplotype> VCFProgenyPhased::collect_haplotypes_mat(
											size_t sample_index) const {
	int side = (this->selection + sample_index) % 2;
	return this->collect_haplotypes_from_phased_progeny(side);
}
vector<Haplotype> VCFProgenyPhased::collect_haplotypes_pat(
											size_t sample_index) const {
	return this->collect_haplotype_from_refs();
}

size_t VCFProgenyPhased::collect_known_parents_indices() const {
	if(this->get_samples()[0] != "0")
		return 0;
	else
		return 1;
}

vector<HaplotypePair> VCFProgenyPhased::impute_core(
									const vector<VCFProgenyPhased *>& vcf_cMs) {
	const size_t	sample_index = vcf_cMs[0]->collect_known_parents_indices();
	
	vector<HaplotypePair>	whole_haplotype_pair = vector<HaplotypePair>();
	const Haplotype	h = Haplotype::default_value();
	HaplotypePair	hap = make_pair(h, h);
	HaplotypePair	*ptr_hap = &hap;
	for(auto p = vcf_cMs.begin(); p != vcf_cMs.end(); ++p) {
		VCFProgenyPhased	*vcf_cM = *p;
		auto	hap_ = vcf_cM->impute_cM_each_sample(*ptr_hap, sample_index,
																false, true);
		whole_haplotype_pair.push_back(hap_);
		ptr_hap = &whole_haplotype_pair.back();
	}
	return whole_haplotype_pair;
}

int VCFProgenyPhased::score_whole(const vector<HaplotypePair>& haps,
							size_t sample_index,
							const vector<VCFProgenyPhased *>& vcf_cMs) const {
	int	score = 0;
	const vector<int>	int_gts = this->get_int_gts(sample_index);
	size_t	first = 0;
	for(size_t i = 0; i < vcf_cMs.size(); ++i) {
		const VCFProgenyPhased *vcf_cM = vcf_cMs[i];
		const HaplotypePair&	hap = haps[i];
		const size_t	last = first + vcf_cM->size();
		const vector<int>	sub_int_gts(int_gts.begin() + first,
										int_gts.begin() + last);
		score += Haplotype::score(hap, sub_int_gts);
		first = last;
	}
	return score;
}

int VCFProgenyPhased::sum_score(const vector<HaplotypePair>& hap_pairs,
								const vector<VCFProgenyPhased *>& vcf_cMs,
								const size_t sample_index) const {
	return this->score_whole(hap_pairs, sample_index, vcf_cMs);
}

void VCFProgenyPhased::impute() {
	const auto	vcf_cMs = VCFImputable::divide_by_cM(this);
	const vector<HaplotypePair>	hap_pairs1 = this->impute_core(vcf_cMs);
	for(auto p = vcf_cMs.begin(); p != vcf_cMs.end(); ++p)
		(*p)->inverse_selection();
	const vector<HaplotypePair>	hap_pairs2 = this->impute_core(vcf_cMs);
	
	const size_t	sample_index = vcf_cMs[0]->collect_known_parents_indices();
	const int	score1 = sum_score(hap_pairs1, vcf_cMs, sample_index);
	const int	score2 = sum_score(hap_pairs2, vcf_cMs, sample_index);
	const auto	hap_pairs = score1 >= score2 ? hap_pairs1 : hap_pairs2;
	for(size_t j = 0; j < vcf_cMs.size(); ++j) {
		vcf_cMs[j]->set_haplotype(hap_pairs[j], sample_index, true);
		vcf_cMs[j]->records.clear();
		delete vcf_cMs[j]->ref_vcf;
		delete vcf_cMs[j];
	}
}

VCFProgenyPhased *VCFProgenyPhased::impute_by_progeny(const VCFSmall *orig_vcf,
													const VCFSmall *imputed_vcf,
													const STRVEC& samples,
													const size_t& ppi,
													const Map& map_,
													const VCFSmall *ref_vcf) {
	VCFProgenyPhased	*vcf = create(orig_vcf, imputed_vcf, samples,
														ppi, map_, ref_vcf);
	vcf->impute();
	return vcf;
}

void VCFProgenyPhased::impute_in_thread(void *config) {
	const auto	*c = (ConfigThread *)config;
	for(size_t i = c->first; i < c->size(); i += c->num_threads) {
		const Family	*family = c->families[i].first;
		const size_t	ppi = c->families[i].second;
		c->results[i] = impute_by_progeny(c->orig_vcf, c->merged_vcf,
												family->get_samples(), ppi,
												c->map, c->ref_vcf);
	}
}

vector<VCFProgenyPhased *> VCFProgenyPhased::impute_all_by_progeny(
						const VCFSmall *orig_vcf, const VCFSmall *merged_vcf,
						const vector<PAIR>& families, const Map& gmap,
						const VCFSmall *ref_vcf, int num_threads) {
	vector<VCFProgenyPhased *>	results(families.size());
	
	const int	T = min((int)families.size(), num_threads);
	vector<ConfigThread *>	configs(T);
	for(int i = 0; i < T; ++i)
		configs[i] = new ConfigThread(orig_vcf, merged_vcf, families,
										gmap, ref_vcf, (size_t)i, T, results);
	
#ifndef DEBUG
	vector<pthread_t>	threads_t(T);
	for(int i = 0; i < T; ++i)
		pthread_create(&threads_t[i], NULL,
						(void *(*)(void *))&impute_in_thread,
						(void *)configs[i]);
	
	for(int i = 0; i < T; ++i)
		pthread_join(threads_t[i], NULL);
#else
	for(int i = 0; i < T; ++i)
		impute_in_thread(configs[i]);
#endif
	
	Common::delete_all(configs);
	return results;
}

VCFProgenyPhased *VCFProgenyPhased::create(const VCFSmall *orig_vcf,
							const VCFSmall *merged_vcf,
							const STRVEC& family_samples, size_t phased_index,
							const Map& map_, const VCFSmall *ref_vcf) {
	// extract parents and the phased progeny
	const STRVEC	samples { family_samples[0], family_samples[1],
												family_samples[phased_index] };
	VCFFamily	*vcf = VCFFamily::create_by_two_vcfs(merged_vcf,
														orig_vcf, samples);
	auto	vcf2 = new VCFProgenyPhased(vcf->get_header(), vcf->get_samples(),
									vcf->get_family_records(), map_, ref_vcf);
	vcf->clear_records();
	delete vcf;
	return vcf2;
}

File: python/VCFProgenyPhased.py
from __future__ import annotations

# coding: utf-8
# VCFProgenyPhased.py
# 子どもが一つでもphasingされた家系
# 親だけphasingする

from collections import defaultdict, Counter
from typing import List, Tuple, Optional, IO, Dict, Iterator

from VCFFamily import *
from VCFImputable import *
from Map import *
import Imputer


#################### VCFProgenyPhased ####################

class VCFProgenyPhased(VCFBase, VCFFamilyBase, VCFImputable):
	def __init__(self, header: list[list[str]],
							records: list[VCFFamilyRecord],
							map_: Map, ref_vcf: VCFSmall):
		VCFBase.__init__(self, header)
		VCFFamilyBase.__init__(self)
		VCFImputable.__init__(self, map_)
		self.records: list[VCFFamilyRecord] = records
		self.ref_vcf: VCFSmall = ref_vcf
		self.selection = 0
	
	def get_header(self) -> list[list[str]]:
		return self.header
	
	def get_samples(self) -> list[str]:
		return self.samples
	
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> VCFRecord:
		return self.records[i]
	
	def get_family_record(self, i: int) -> VCFFamilyRecord:
		return self.records[i]
	
	def clip_ref_haplotype(self, sample_id: int, side: int) -> Haplotype:
		hap = self.ref_vcf.clip_raw_haplotype(sample_id, side)
		return Haplotype(hap, sample_id, side)
	
	# divide VCF by 1 cM
	def divide_by_cM(self) -> Iterator[VCFProgenyPhased]:
		# 1cMを超えても10個以内で10cM以内なら塊とみなす
		rs = self.records[:1]
		ref_rs = self.ref_vcf.records[:1]
		for i in range(1, len(self)):
			record = self.records[i]
			ref_record = self.ref_vcf.records[i]
			if self.is_block(record, rs):
				rs.append(record)
				ref_rs.append(ref_record)
			else:
				ref_vcf_cM = VCFSmall(self.ref_vcf.header, ref_rs)
				yield VCFProgenyPhased(self.header, rs, self.map, ref_vcf_cM)
				rs = [record]
				ref_rs = [ref_record]
		
		ref_vcf_cM = VCFSmall(self.ref_vcf.header, ref_rs)
		yield VCFProgenyPhased(self.header, rs, self.map, ref_vcf_cM)
	
	def collect_haplotypes_from_phased_progeny(self, j: int) -> list[Haplotype]:
		return [ self.clip_haplotype(2, j) ]
	
	def collect_haplotype_from_refs(self) -> list[Haplotype]:
		return [ self.clip_ref_haplotype(i, j)
					for i in range(self.ref_vcf.num_samples())
					for j in (0, 1) ]
	
	def collect_haplotypes_mat(self, sample_index: int) -> list[Haplotype]:
		i = (self.selection + sample_index) % 2
		return self.collect_haplotypes_from_phased_progeny(i)
	
	def collect_haplotypes_pat(self, sample_index: int) -> list[Haplotype]:
		return self.collect_haplotype_from_refs()
	
	def impute_cM(self, prev_haps: list[HaplotypePair],
							sample_indices: list[int]) -> list[HaplotypePair]:
		haps: list[HaplotypePair] = []
		for sample_index, prev_hap in zip(sample_indices, prev_haps):
			hap = self.impute_cM_each_sample(prev_hap, sample_index, False)
			haps.append(hap)
		return haps
	
	def impute_core(self, vcf_cMs: list[VCFProgenyPhased]
											) -> list[list[HaplotypePair]]:
		sample_indices = [ i for i, s in enumerate(vcf_cMs[0].samples[:2])
																if s != '0' ]
		N = len(sample_indices)
		whole_haplotype_pair: list[list[HaplotypePair]] = [[] for _ in range(N)]
		h = Haplotype.default_value()
		haps = [(h, h)] * len(sample_indices)
		for vcf_cM in vcf_cMs:
			haps = vcf_cM.impute_cM(haps, sample_indices)
			for i in range(N):
				whole_haplotype_pair[i].append(haps[i])
		return whole_haplotype_pair
	
	def score_whole(self, haps: list[HaplotypePair],
					sample_index: int, vcf_cMs: list[VCFProgenyPhased]) -> int:
		score = 0
		int_gts = self.get_int_gts(sample_index)
		first = 0
		for i in range(len(vcf_cMs)):
			vcf_cM = vcf_cMs[i]
			hap = haps[i]
			last = first + len(vcf_cM)
			score += Haplotype.score(int_gts[first:last], hap[0], hap[1])
			first = last
		return score
	
	def inverse_selection(self):
		self.selection = 1 if self.selection == 0 else 0
	
	def impute(self):
		vcf_cMs = list(self.divide_by_cM())
		hap_pairs1 = self.impute_core(vcf_cMs)
		for vcf_cM in vcf_cMs:
			vcf_cM.inverse_selection()
		hap_pairs2 = self.impute_core(vcf_cMs)
		sample_indices = [ i for i, s in enumerate(vcf_cMs[0].samples[:2])
																if s != '0' ]
		score1 = sum(self.score_whole(hap_pairs1[i], i, vcf_cMs)
												for i in sample_indices)
		score2 = sum(self.score_whole(hap_pairs2[i], i, vcf_cMs)
												for i in sample_indices)
		_, hap_pairs = max([(score1, hap_pairs1), (score2, hap_pairs2)],
														key=lambda v: v[0])
		for j, vcf_cM in enumerate(vcf_cMs):
			for i in sample_indices:
				vcf_cM.set_haplotype(hap_pairs[i][j], i)
	
	@staticmethod
	def create(orig_vcf: VCFSmall, merged_vcf: VCFSmall,
							family_samples: list[str], phased_index: int,
							map_: Map, ref_vcf: VCFSmall) -> VCFProgenyPhased:
		# extract parents and the phased progeny
		samples = family_samples[:2] + [family_samples[phased_index]]
		vcf = VCFFamily.create_by_two_vcfs(merged_vcf, orig_vcf, samples)
		return VCFProgenyPhased(vcf.header, vcf.records, map_, ref_vcf)

__all__ = ['VCFProgenyPhased']
